### PR TITLE
add validation for password and username

### PR DESCRIPTION
**Pull request by jonathaneeckhout**
New:
- username validation
- password validation

Fix: 
- Moved the validation part to the general database function instead of the backend

Closes https://github.com/jonathaneeckhout/jdungeon/issues/36 

How to test:
- try to create empty string users -> should fail
- try to create user with a short username, a username with a special character or longer than 16 characters -> should fail
- try to create a short password, a password with spaces or longer than 32 characters -> should fail

### CODE REVIEW

**Comment by ivoryfitz** [comment]
- try to create empty string users -> should fail
  - ✅ 
- try to create user with a short username, a username with a special character or longer than 16 characters -> should fail
  - ✅ 
- try to create a short password, as password with spaces or longer than 32 characters -> should fail
  -  ✅

File: scripts/network/database/database.gd [hunk 1]
```diff
@@ -26,13 +26,41 @@ func init() -> bool:
...
 func create_account(username: String, password: String) -> bool:
-	if username == "" or password == "":
+	if username == "" or password == "" or not is_account_valid(username, password):
 		J.logger.info("Invalid username or password")
 		return false
 
 	return backend.CreateAccount(username, password)
 
 
+func is_account_valid(username: String, password: String) -> bool:
```
Comment (ivoryfitz) [inline scripts/network/database/database.gd:36]: I think we will want a way at some point to bubble up the exact error messages to inform the user as to why their account creation failed. Can do here, or pop a new issue and address later. 

A possible solution could be:

```
func is_account_valid(username: String, password: String) -> Dictionary:
  ...
  if username.length() < 4 or username.length() > 16:
    var validation_error_message := "Username must be at least 4 characters long or maximum 16 characters long."
    J.logger.warn(validation_error_message)
    return {"validation_status": false, "validation_error_message": validation_error_message}
  ...
```

Or there may be some other solution. I wish GDscript supported multi-value returns, but from what I see it does not. It'd be nice to do

```
func foo() -> (ret_val, message)

var return_val, message = foo()
```

But I don't think we have that, so for multi-var returns its either an array or dictionary.

**Review by ivoryfitz — APPROVED** [review]
Looks fine to me, though up to you what you'd like to do with bubbling the validation messages back out to the UI so the user can know what's wrong.

**Comment by jonathaneeckhout** [comment]
> Looks fine to me, though up to you what you'd like to do with bubbling the validation messages back out to the UI so the user can know what's wrong.

It's a valid remark, but I want to keep the server side as short and light as possible. If it fails, too bad: false.

How I've seen it in other projects is that they do client side validation as well. Which gives more information to the user.

**Comment by jonathaneeckhout** [comment]
> > Looks fine to me, though up to you what you'd like to do with bubbling the validation messages back out to the UI so the user can know what's wrong.
> 
> It's a valid remark, but I want to keep the server side as short and light as possible. If it fails, too bad: false.
> 
> How I've seen it in other projects is that they do client side validation as well. Which gives more information to the user.

Nevermind I will implement the reason, thx for the comments

**Comment by ivoryfitz** [comment]
I'm fine with either, as long as at _some_ point we get a better error message back to the user so they can know why they failed to create, I'm good with it.

**Comment by jonathaneeckhout** [comment]
> I'm fine with either, as long as at _some_ point we get a better error message back to the user so they can know why they failed to create, I'm good with it.

I've modified the code to provide feedback, the only case which isn't covered is when a player already exists, it will just say: Oops something went wrong. But this one is going to be something for a future PR.

**Pull request merged**